210 add create and delete functionality

tulza opened the pull request:
closes #210 

- create and delete dialog
- created dialog UI for create and delete
- modify semester api to handle name error
- modify navigation bar
- added back button component
- modify textInput
- style fixes directly in shadcn component

tulza committed:
Add dialog for create and delete

- created create and delete dialog
- modify text input to handle placeholder correctly
- modify dialogUtils for customisability

tulza committed:
Add create and delete fuctionality + Refactors

- Added create and delete functionality
- Refactor optionsDialog into smaller components

tulza committed:
Improve error handling, loading state and api

- added a reponse in the api to check if name exist
- refactor dialogs to use response body on error message
- added loading state on dialog footer

tulza commented:
oops

tulza committed:
minor fixes

tulza committed:
Merge branch 'main' into Uabc-210-add-create-and-delete-functionality

tulza committed:
Add create semester button

dyzhuu commented on src/components/admin/semesters/SemesterCreateDialog.tsx:
```diff
@@ -0,0 +1,187 @@
+      .refine(validateDate, "Invalid date"),
+  })
+  .refine((data) => compareDate(data.startDate, data.breakStart) < 0, {
+    message: "Start date must be before than break start date",
```
check grammar here ane below as well: `Start date must be before break start date`

dyzhuu replied:
i suppose this also means that you should change the error messages in SemesterEditDialog? Good to have consistency. (`before` instead of `less than`)

tulza replied:
oops

dyzhuu commented on src/components/admin/semesters/SemesterDeleteDialog.tsx:
```diff
@@ -1,17 +1,77 @@
+        method: "DELETE",
+        headers: {
+          "Content-Type": "application/json",
+        },
```
no need for headers.

dyzhuu commented on src/components/admin/semesters/SemesterDeleteDialog.tsx:
```diff
@@ -1,17 +1,77 @@
-      <DialogCardFooter variant="destructive" />
+      <div className="bg-destructive/20 w-full p-4 rounded-lg select-none">
+        <p className="font-bold text-destructive">Warning</p>
+        <p className="text-sm text-destructive">This action is irreversible.</p>
```
oh maybe mention that all related gameSessionSchedules will be deleted?

dyzhuu commented on src/app/api/semesters/route.ts:
```diff
@@ -62,6 +63,18 @@ export async function POST(req: NextRequest) {
       });
     }
 
+    const findName = await db
+      .select()
+      .from(semesters)
+      .where(eq(semesters.name, newSemester.name))
+      .limit(1);
+
+    if (findName.length != 0) {
+      return new Response("This name already exists, please pick another", {
+        status: 400,
+      });
+    }
+
```
rename this to something better like `existingSemester` please. This gives more context to what the variable actually is and allows follow up code to be more readable as well. e.g.

```
if (!existingSemester) {
    ...
}
```

to get a single semester instead of a array you could destructure it like so

```
const [existingSemester] = await db.select().....
```

or use the `findFirst` function provided by the query api

```
const existingSemester = await db.query.semesters.findFirst({
    where: ...
})
```

dyzhuu commented on src/components/ui/utils/DialogUtils.tsx:
```diff
@@ -29,15 +36,78 @@ export const DialogCard = ({ title, children, onClose }: DialogCardProps) => {
+      <Dialog
+        onOpenChange={() => {
+          setOpen(!open);
+          onOpenChange && onOpenChange();
```
is it necessary to check if it exists? maybe just set a default void function to onOpenChange?

dyzhuu commented on src/components/ui/utils/DialogUtils.tsx:
```diff
@@ -29,15 +36,78 @@ export const DialogCard = ({ title, children, onClose }: DialogCardProps) => {
-      <Button {...props}>Confirm</Button>
+      <Button
+        {...props}
+        className={clsx(
```
can we use cn instead of clsx

dyzhuu commented on src/components/admin/semesters/SemesterCreateDialog.tsx:
```diff
@@ -0,0 +1,187 @@
+      breakEnd: formatDateInISO(data.breakEnd),
+      bookingOpenDay: "Monday",
+      bookingOpenTime: "12:00:00",
+      createAt: "2022-01-01T00:00:00.000Z",
+    });
```
the `createdAt` field is autogenerated, no need to provide

dyzhuu commented on src/components/admin/semesters/SemesterCreateDialog.tsx:
```diff
@@ -0,0 +1,187 @@
+        queryClient.invalidateQueries({ queryKey: ["semesters"] });
+        toast({
+          title: "Success!",
+          description: "Semester details successfully updated",
+        });
```
should say `successfully created`

dyzhuu commented on src/components/admin/semesters/SemesterCreateDialog.tsx:
```diff
@@ -0,0 +1,187 @@
+      });
+
+      if (!response.ok) {
+        await response.text().then((text) => {
+          // todo: handle this better somehow
+          if (text == "This name already exists, please pick another") {
+            setError("name", { message: text });
+          }
+          throw new Error(text || "An has error occurred");
+        });
```
you could consider attaching something to the error message to detect that the error is due to a duplicate semester name

e.g. return a json with errorCode field that identifies the type of error

i want to eventually overhaul how errors are returned from the backend to make things more detailed too.

dyzhuu commented on src/components/admin/semesters/SemesterCreateDialog.tsx:
```diff
@@ -0,0 +1,187 @@
+  });
+  const { toast } = useToast();
+
+  // React-query
```
not sure about the necessity of these comments

dyzhuu commented on src/components/admin/semesters/SemesterDeleteDialog.tsx:
```diff
@@ -1,17 +1,77 @@
+        <p className="font-bold text-destructive">Warning</p>
+        <p className="text-sm text-destructive">This action is irreversible.</p>
+      </div>
+      <form onSubmit={onSubmit}>
```
this usage of form is fine but we could probably just have an onClick.

dyzhuu commented on src/components/admin/semesters/SemesterDeleteDialog.tsx:
```diff
@@ -1,17 +1,77 @@
+        <DialogCardFooter
+          variant="destructive"
+          primaryText="Delete"
+          secondaryText="No, Cancel"
```
```suggestion
          secondaryText="Cancel"
```

dyzhuu commented on src/components/ui/optionsPopover/OptionsPopover.tsx:
```diff
@@ -47,23 +31,11 @@ export const OptionDialogItem = ({
-      </Dialog>
-    </OptionsDialogContext.Provider>
+    <DialogBase onOpenChange={() => closePopover()}>
+      <DialogTrigger asChild>{ButtonElement}</DialogTrigger>
+      {DialogElement}
```
actually -- can we make this say ButtonComponent and DialogComponent
think its more clear that way

dyzhuu commented on src/components/ui/utils/DialogUtils.tsx:
instead of these helpers, feel free to directly edit the `dialog.tsx` file in `@/components/ui`.

The `DialogBase` component can probably be merged with their `Dialog` component (and you would use `DialogPrimitive.Root` instead of `Dialog`)

Same with the footer.

![image](https://github.com/UoaWDCC/uabc-portal/assets/119282717/07379eb2-2312-424f-89f6-565b968e65cf)

dyzhuu replied:
and I'm not sure if I like the custom DialogCard component. Think its better to separate the content and the header. If you have styling changes you want to make globally just do it directly in `dialog.tsx.`

tulza committed:
Resolve conversations

- Improved delete dialog warning
- Improved error handling for semesters api
- Renamed optionsPopover params to improve clarity
- Fixed grammars on dialog errors
- Fixed minor bug

tulza committed:
Modify Dialog.tsx

- Provides dialog for all Dialog
- Re-import all DialogBase instance to dialog
- Removed DialogBase

tulza committed:
Remove DialogCard

tulza committed:
Added documentations + minor style changes

tulza committed:
Merge branch 'main' into Uabc-210-add-create-and-delete-functionality
